text/0000-dockerfiles.md

```diff
@@ -0,0 +1,228 @@
+- Author(s): sclevine
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: [RFC0069](https://github.com/buildpacks/rfcs/blob/main/text/0069-stack-buildpacks.md), [RFC#167](https://github.com/buildpacks/rfcs/pull/167)
+- Depends on: [RFC#172](https://github.com/buildpacks/rfcs/pull/172)
+
+# Summary
+[summary]: #summary
+
+This RFC introduces functionality for customizing base images, as an alternative to stackpacks.
+
+# Motivation
+[motivation]: #motivation
+
+Relying on Dockerfiles for base image generation and manipulation allows us to apply buildpacks only to the problem that they solve best: managing application runtimes and dependencies.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes that we replace stackpacks with build-time and runtime Dockerfiles that act as pre-build hooks.
+These hooks would be equivalent to base image / builder customizations, but would execute immediately before a build.
+
+Most importantly: the changes below do not impact buildpacks or the buildpack API.
+For a given application, a build that uses hooks could be optimized by creating a more narrowly-scoped builder with the hooks pre-applied, without modifying the buildpacks.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Note: kaniko, buildah, BuildKit, or the original Docker daemon may be used to apply Dockerfiles at the platform's discretion. 
+
+### Dynamically-applied Dockerfiles
+
+A builder image may include any number of "hook" files in `/cnb/hooks.d/`. Hooks may be associated with zero or more buildpack IDs.
+The `pack create-builder` command is used to copy hooks into the builder image via `builder.toml`. E.g.,
+```toml
+[[hooks]]
+name = "app-hook"
+path = "./myhook"
+buildpacks = ["com.example.buildpack1"]
+```
+
+(Note: this format should not be considered final. Additional fields will be required to mark the format, target, etc. of the hook. These may be defined in subsequent spec PRs / sub-team RFCs.)
+
+Each hook file path must be in the format `/cnb/hooks.d/<name>.(build.|run.|)<format>(.out|)`, where:
+
+1. `build`, `run`, or empty specify the builder image, runtime base image, or both base images, respectively.
+2. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future.
+3. If the `.out` suffix is present and the file is executable, the hook is executed in the context of the app directory, and its output must match `<format>`.
+4. If the `.out` suffix is not present, the contents of the file must match `<format>`.
+
+Additionally, each hook file may be accompanied by an additional file of the same name, but with the suffix `.buildpacks`.
+This file contains a newline-separated list of buildpack IDs.
+If present, the hook is ignored unless one of the buildpacks in the list is selected during the buildpack detection phase.
+
+Hook files are evaluated (either read or executed) after the detection phase.
```

Comment:
Can you include here a section to explain how the hooks will be selected from `/cnb/hooks.d/` ? If a group of hooks to be applied is generated, how will the hooks be sorted, .... ? Will the process rollback what it has been previously executed if a hook of the group fails now ?

Reply:
> If a group of hooks to be applied is generated, how will the hooks be sorted, .... ?

As specified,

> The resulting Dockerfiles are applied to the build-time and runtime base images in lexographical order by `<name>`.
(there are more details around that section: https://github.com/buildpacks/rfcs/pull/173/files#diff-aeaa4e3f3765f6b7ec295a8c92881bdeae1313f69fd338da209548883dafdf00R62)

> Will the process rollback what it has been previously executed if a hook of the group fails now ?

As specified,

> If an executable hook exits with a non-zero status value, the build fails.

Reply:
I assume this is a new phase? Maybe the `extend` phase :)